File: backend/src/main/java/com/kikiwiki/backend/transaction/Transaction.java
package com.kikiwiki.backend.transaction;

import jakarta.persistence.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "transactions")
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // 금액은 실수(double) 대신 BigDecimal을 씀 - 부동소수점 오차로 돈 계산이 틀어지는 걸 방지
    @Column(nullable = false)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;

    @Column(nullable = false)
    private String category;

    private String description;

    @Column(nullable = false)
    private LocalDate transactionDate;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    // JPA는 인자 없는 생성자가 반드시 필요함 (엔티티를 DB에서 읽어올 때 내부적으로 사용)
    protected Transaction() {
    }

    public Transaction(BigDecimal amount, TransactionType type, String category,
                        String description, LocalDate transactionDate) {
        this.amount = amount;
        this.type = type;
        this.category = category;
        this.description = description;
        this.transactionDate = transactionDate;
    }

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }

    public Long getId() {
        return id;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public TransactionType getType() {
        return type;
    }

    public String getCategory() {
        return category;
    }

    public String getDescription() {
        return description;
    }

    public LocalDate getTransactionDate() {
        return transactionDate;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
}

File: frontend/src/types/transaction.ts
export type TransactionType = 'INCOME' | 'EXPENSE';

export interface Category {
  id: number;
  name: string;
  type: TransactionType;
}

export interface Transaction {
  id: number;
  amount: number;
  type: TransactionType;
  categoryId: number;
  categoryName: string;
  description: string | null;
  transactionDate: string; // 'YYYY-MM-DD' 형식
  createdAt: string;
}

export interface TransactionInput {
  amount: number;
  type: TransactionType;
  categoryId: number;
  description: string;
  transactionDate: string;
}

export interface TransactionPage {
  items: Transaction[];
  hasMore: boolean;
  totalCount: number;
}
